---
title: "Mouse hypothalamus data by MERFISH (Moffitt et al., 2018)"
author: "Zheng Li"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  workflowr::wflow_htlm:
    toc: true
editor_options:
  chunk_output_type: console
---

# Introduction
Here, we apply BASS to analyze the spatial transcriptomic data that measured the 
mouse preoptic region of hypothalamus by the MERFISH technology from 
[Moffitt et al., 2018](https://www.science.org/doi/10.1126/science.aau5324). We 
focus on the tissue sections Bregma -0.04, -0.09, -0.14, -0.19, and -0.24 mm 
from a consecutive brain hypothalamic region of animal 1. The original data can 
be downloaded from 
[here](https://datadryad.org/stash/dataset/doi:10.5061/dryad.8t8s248). 
We excluded five blank control genes and excluded cells that were annotated to 
be the "Ambiguous" class as they were identified as putative doublets. Finally, 
we retained the same set of 155 genes measured on 5,488 (Bregma -0.04 mm), 
5,557 (Bregma -0.09 mm), 5,926 (Bregma -0.14 mm), 5,803 (Bregma -0.19 mm), and 
5,543 (Bregma -0.24 mm) cells along with their centroid coordiates for the 
downstream analysis. We manually annotated the five samples with spatial domain 
labels, allowing us to quantitatively evaluate different methods with ARI. 
There, the manual annotation of spatial domains was based on the spatial 
expression of marker genes and the histology diagram of the mouse brain from the 
[Allen brain atlas](https://mouse.brain-map.org/static/atlas). The processed 
data and manual annotations can be found at the 
[data directory](https://github.com/zhengli09/BASS-Analysis/tree/master/data). 
For detailed usage of all the functions, please refer to the software tutorial 
section.

```{r error=FALSE,message=FALSE}
library(BASS)
library(Seurat)
library(tidyverse)
load("data/MERFISH_Animal1.RData") # cnts_mult info_mult
```

# Load data and set hyper-parameters
```{r}
smps <- c("-0.04", "-0.09", "-0.14", "-0.19", "-0.24")
cnts <- cnts_mult[smps] # a list of gene expression count matrices
xys <- lapply(info_mult[smps], function(info.i){
  info.i$x <- info.i$x - min(info.i$x)
  info.i$y <- info.i$y - min(info.i$y)
  as.matrix(info.i[, c("x", "y")])
}) # a list of spatial coordinates matrices
# hyper-parameters
C <- 20 # number of cell types
R <- 8 # number of spatial domains
```

# Run BASS
```{r warning=FALSE}
set.seed(0)
# Set up BASS object
BASS <- createBASSObject(cnts, xys, C = C, R = R, beta_method = "SW")

# Data pre-processing:
# 1.Library size normalization followed with a log2 transformation
# 2.Dimension reduction with PCA after standardizing all the genes
# 3.Batch effect adjustment using the Harmony package
BASS <- BASS.preprocess(BASS, doLogNormalize = TRUE,
  doPCA = TRUE, scaleFeature = TRUE, nPC = 20)

# Run BASS algorithm
BASS <- BASS.run(BASS)
```
```{r fig.height=4.5, fig.width=4.5}
# The spatial parameter beta has converged
# after checking the trace plot of beta
plot(1:BASS@burnin, BASS@samples$beta, xlab = "Iteration", 
  ylab = expression(beta), type = "l")
```

```{r}
# post-process posterior samples:
# 1.Adjust for label switching with the ECR-1 algorithm
# 2.Summarize the posterior samples to obtain the cell type labels, spatial 
#   domain labels, and cell type proportion matrix estimate
BASS <- BASS.postprocess(BASS)
clabels <- BASS@results$c # cell type clusters
zlabels <- BASS@results$z # spatial domain labels
pi_est <- BASS@results$pi # cell type composition matrix
```

# Annotate cell types
```{r}
# Perform DE analysis with Seurat
cnts_all <- do.call(cbind, cnts)
seu_obj <- CreateSeuratObject(counts = cnts_all)
seu_obj <- NormalizeData(seu_obj)
seu_obj <- ScaleData(seu_obj, features = rownames(seu_obj))
seu_obj <- RunPCA(seu_obj, features = rownames(seu_obj), verbose = F)
Idents(seu_obj) <- factor(unlist(clabels))
markers <- FindAllMarkers(seu_obj, only.pos = T, 
  min.pct = 0, logfc.threshold = 0, verbose = F)
top5 <- markers %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_logFC)

# By checking the top DE genes of each cell type cluster, 
# we annotate specific cell types for each cluster
cTypes <- c(
  "Astro", "Epen", "E5", "I7", "Endo",
  "I6", "E2", "I1", "Mural", "I4",
  "I3", "I8", "I5", "E1", "E4",
  "I2", "MOD", "IOD", "E3", "Micro")
clabels <- lapply(clabels, function(clabels.l){
  clabels.l <- factor(clabels.l)
  levels(clabels.l) <- cTypes
  clabels.l <- factor(clabels.l, levels = c(
    "E1", "E2", "E3", "E4", "E5",
    "I1", "I2", "I3", "I4", "I5", 
    "I6", "I7", "I8", "MOD", "IOD", 
    "Astro", "Micro", "Epen", "Endo", "Mural"
  ))
})
```

<p>
  <a class="btn btn-primary" data-toggle="collapse" href="#collapseExample1" role="button" aria-expanded="false" aria-controls="collapseExample1">
    Top DE genes for each cell type cluster
  </a>
</p>
<div class="collapse" id="collapseExample1">
  <div class="card card-body">

  ```{r}
  data.frame(top5)
  ```

  </div>
</div>

# Visualization
You can refer to 
[visualization](https://github.com/zhengli09/BASS-Analysis/tree/master/code/viz.R) 
for some useful plotting functions or you can write your own code for plotting.
```{r}
source("code/viz.R")
```

## Spatial domains
```{r fig.height=3, fig.width=12}
zlabels <- lapply(zlabels, function(zlabels.l){
  zlabels.l <- factor(zlabels.l)
  levels(zlabels.l) <- c("BST", "PV", "PVT", "MPN",
    "V3", "fx", "MPA", "PVH")
  zlabels.l <- factor(zlabels.l, levels = c("V3", "BST",
    "fx", "MPA", "MPN", "PV", "PVH", "PVT"))
})
cols <- c("#669bbc", "#81b29a", "#f2cc8f", "#adc178",
  "#dde5b6", "#a8dadc", "#e5989b", "#e07a5f")
plotClustersFacet(xys, zlabels, smps, size = 0.2) +
  scale_color_manual("Domain", values = cols) +
  guides(color = guide_legend(nrow = 1, 
    override.aes = list(size = 2)))
```

## Cell types
```{r fig.height=3.5, fig.width=12}
cols <- c("#98c1d9", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51", 
  "#e0fbfc", "#a8dadc", "#3d5a80", "#81b29a", "#e07a5f", 
  "#DBC9D8", "#b388eb", "#A4277C", "#BC93B2", "#0077b6", 
  "#bb3e03", "#ffddd2", "#f19c79", "#006d77", "#6A3569")
plotClustersFacet(xys, clabels, smps, size = 0.2) +
  scale_color_manual("Type", values = cols) +
  guides(color = guide_legend(byrow = T, nrow = 2,
    override.aes = list(size = 2)))
```

## Spatial distribution of cell types on the tissue section Bregma -0.14
```{r message=FALSE, warning=FALSE, fig.height=7, fig.width=7}
library(gghighlight)
plotCellTypes(xys[[3]], clabels[[3]], cols, ncol = 5, dotsize = 0.5,
  size = 0.6, alpha = 0.5, stroke = 0)
```

## Cell type compositions in each spatial domain
```{r fig.height=4, fig.width=5}
# adjust order of labeling
pi_est <- pi_est[match(
  levels(clabels[[1]]), cTypes),
  c(5,1,6,7,4,2,8,3)]
colnames(pi_est) <- c("V3", "BST", "fx", "MPA", "MPN", "PV", "PVH", "PVT")
rownames(pi_est) <- levels(clabels[[1]])
plotCellTypeCompBar(pi_est, cols, nrow = 4)
```

